File: src/main/java/me/miki/shindo/ui/comp/addons/CompComboBox.java
package me.miki.shindo.ui.comp.addons;

import lombok.Getter;
import me.miki.shindo.Shindo;
import me.miki.shindo.management.addons.settings.impl.ComboSetting;
import me.miki.shindo.management.color.AccentColor;
import me.miki.shindo.management.color.ColorManager;
import me.miki.shindo.management.nanovg.NanoVGManager;
import me.miki.shindo.management.nanovg.font.Fonts;
import me.miki.shindo.ui.comp.Comp;
import me.miki.shindo.utils.MathUtils;
import me.miki.shindo.utils.animation.simple.SimpleAnimation;
import me.miki.shindo.utils.mouse.MouseUtils;

import java.awt.*;

public class CompComboBox extends Comp {

    private final SimpleAnimation changeAnimation = new SimpleAnimation();
    @Getter
    private final ComboSetting setting;
    private final float width;
    private int changeDirection;

    public CompComboBox(float x, float y, float width, ComboSetting setting) {
        super(x, y);
        this.width = width;
        this.setting = setting;
        this.changeDirection = 1;
        this.changeAnimation.setValue(1);
    }

    public CompComboBox(float width, ComboSetting setting) {
        super(0, 0);
        this.width = width;
        this.setting = setting;
        this.changeDirection = 1;
        this.changeAnimation.setValue(1);
    }

    @Override
    public void draw(int mouseX, int mouseY, float partialTicks) {

        Shindo instance = Shindo.getInstance();
        NanoVGManager nvg = instance.getNanoVGManager();
        ColorManager colorManager = instance.getColorManager();
        AccentColor accentColor = colorManager.getCurrentColor();

        changeAnimation.setAnimation(changeDirection, 16);

        nvg.drawGradientRoundedRect(this.getX(), this.getY(), width, 16, 4, accentColor.getColor1(), accentColor.getColor2());

        nvg.drawCenteredText(setting.getOption().getName(), this.getX() + (width / 2) + ((changeDirection - changeAnimation.getValue()) * 22), this.getY() + 5F, new Color(255, 255, 255, (int) (MathUtils.abs(changeAnimation.getValue() * 255))), 8, Fonts.REGULAR);

        nvg.drawText("<", this.getX() + 4, this.getY() + 4F, Color.WHITE, 10, Fonts.REGULAR);
        nvg.drawText(">", this.getX() + width - 10, this.getY() + 4F, Color.WHITE, 10, Fonts.REGULAR);
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {

        int max = setting.getOptions().size();
        int modeIndex = setting.getOptions().indexOf(setting.getOption());

        if (mouseButton == 0) {

            if (MouseUtils.isInside(mouseX, mouseY, this.getX(), this.getY(), 16, 16)) {

                changeAnimation.setValue(0);

                if (modeIndex > 0) {
                    modeIndex--;
                } else {
                    modeIndex = max - 1;
                }

                changeDirection = 1;
                setting.setOption(setting.getOptions().get(modeIndex));
            }

            if (MouseUtils.isInside(mouseX, mouseY, this.getX() + width - 16, this.getY(), 16, 16)) {

                changeAnimation.setValue(0);

                if (modeIndex < max - 1) {
                    modeIndex++;
                } else {
                    modeIndex = 0;
                }

                changeDirection = -1;
                setting.setOption(setting.getOptions().get(modeIndex));
            }
        }
    }

}

File: src/main/kotlin/me/miki/shindo/ui/components/v2/inputs/CompComboBox.kt
package me.miki.shindo.ui.components.v2.inputs

import me.miki.shindo.management.nanovg.font.Fonts
import me.miki.shindo.management.settings.impl.ComboSetting
import me.miki.shindo.ui.animation.v2.value.SimpleAnimation
import me.miki.shindo.ui.components.v2.Component
import me.miki.shindo.utils.MathUtils
import me.miki.shindo.utils.mouse.MouseUtils
import java.awt.Color


class CompComboBox : Component {
    private val changeAnimation = SimpleAnimation()
    private val setting: ComboSetting
    private var width: Float
    private var changeDirection: Int

    constructor(x: Float, y: Float, width: Float, setting: ComboSetting) : super(x, y) {
        this.width = width
        this.setting = setting
        this.changeDirection = 1
        this.changeAnimation.setValue(1f)
        setWidth(width)
        super.setHeight(CONTROL_HEIGHT)
    }

    constructor(width: Float, setting: ComboSetting) : super(0f, 0f) {
        this.width = width
        this.setting = setting
        this.changeDirection = 1
        this.changeAnimation.setValue(1f)
        setWidth(width)
        super.setHeight(CONTROL_HEIGHT)
    }

    override fun getWidth(): Float = width

    override fun setWidth(width: Float) {
        this.width = width
        super.setWidth(width)
    }

    override fun draw(mouseX: Int, mouseY: Int, partialTicks: Float) {
        val nvgInstance = nvg
        val accentColor = accent

        changeAnimation.setAnimation(changeDirection.toFloat(), 16.0)

        nvgInstance.drawGradientRoundedRect(
            this.getX(),
            this.getY(),
            width,
            16f,
            4f,
            accentColor.getColor1(),
            accentColor.getColor2()
        )

        nvgInstance.drawCenteredText(
            setting.getOption()!!.name,
            this.getX() + (width / 2f) + ((changeDirection - changeAnimation.getValue()) * 22f),
            this.getY() + 5f,
            Color(255, 255, 255, (MathUtils.abs(changeAnimation.getValue() * 255)).toInt()),
            8f,
            Fonts.REGULAR
        )

        nvgInstance.drawText("<", this.getX() + 4, this.getY() + 4f, Color.WHITE, 10f, Fonts.REGULAR)
        nvgInstance.drawText(">", this.getX() + width - 10, this.getY() + 4f, Color.WHITE, 10f, Fonts.REGULAR)
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        val max = setting.getOptions().size
        var modeIndex = setting.getOptions().indexOf(setting.getOption())

        if (mouseButton == 0) {
            if (MouseUtils.isInside(mouseX, mouseY, this.getX(), this.getY(), 16f, 16f)) {
                changeAnimation.setValue(0f)

                if (modeIndex > 0) {
                    modeIndex--
                } else {
                    modeIndex = max - 1
                }

                changeDirection = 1
                setting.setOption(setting.getOptions()[modeIndex])
            }

            if (MouseUtils.isInside(mouseX, mouseY, this.getX() + width - 16f, this.getY(), 16f, 16f)) {
                changeAnimation.setValue(0f)

                if (modeIndex < max - 1) {
                    modeIndex++
                } else {
                    modeIndex = 0
                }

                changeDirection = -1
                setting.setOption(setting.getOptions()[modeIndex])
            }
        }

        super.mouseClicked(mouseX, mouseY, mouseButton)
    }

    companion object {
        private const val CONTROL_HEIGHT = 16f
        private const val BUTTON_SIZE = 16f
    }
}
